controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -259,6 +268,41 @@ func (r *BareMetalHostReconciler) Reconcile(request ctrl.Request) (result ctrl.R
 	return
 }
 
+// Consume inspect.metal3.io/hardwaredetails when either
+// inspect.metal3.io=disabled or there are no existing HardwareDetails
+func (r *BareMetalHostReconciler) updateHardwareDetails(request ctrl.Request, host *metal3v1alpha1.BareMetalHost) (bool, error) {
+	reqLogger := r.Log.WithValues("baremetalhost", request.NamespacedName)
+	updated := false
+	if host.Status.HardwareDetails == nil || inspectionDisabled(host) {
+		objHardwareDetails, err := r.getHardwareDetailsFromAnnotation(host)
+		if err != nil {
+			reqLogger.Info("Error getting HardwareDetails from annotation")
+			return updated, err
+		} else if objHardwareDetails != nil {
+			reqLogger.Info("Setting HardwareDetails from annotation")
+			host.Status.HardwareDetails = objHardwareDetails
+			err = r.saveHostStatus(host)
+			if err != nil {
+				reqLogger.Info("Could not update hardwaredetails from annotation")
+				return updated, err
+			}
+			updated = true
+		}
+	}
+	// We either just processed the annotation, or the status is already set
+	// so we remove it
+	annotations := host.GetAnnotations()
+	if _, present := annotations[hardwareDetailsAnnotation]; present {
+		reqLogger.Info("Removing HardwareDetails annotation")
+		delete(host.Annotations, hardwareDetailsAnnotation)
+		err := r.Update(context.TODO(), host)
```

Comment:
If we save the status above, will the host be out of date and trigger an error updating it's annotations here? I remember seeing those sorts of errors early on in development, but it may have been from multiple updates to the same part of the resource or some other condition, so maybe it isn't an issue here.

Reply:
I tested locally and this seems to work, but I wasn't sure since the status annotation code does return from the Reconcile() after each update - I could do the same here if we think that's safer?

Part of the problem is the unit tests don't exhibit the same behavior wrt resource updates that we see on the live cluster (we just look at the modified host object, not the stored state), so it's more difficult to be sure we get this kind of thing right.

Reply:
Agreed, we need to figure out a way to add end-to-end tests for some of these more complicated behaviors.

Reply:
I think it should be ok, and agreed that we need to find a better approach to such kind of tests